imap_processing/lo/l0/science_direct_events.py

```diff
@@ -85,116 +85,155 @@ class ScienceDirectEvents(LoL0):
             # We only need the last 12 bits from the tof calculation binary
-            tof_calculation_array = bitarray(tof_calculation_binary).tolist()[-12:]
+            tof_calculation_array = bitarray(tof_binary).tolist()[-12:]
 
             # the tof calculation table binary are used to determine which values
             # from the tof coefficient table should be used in combination with
             # the tof decoder.
-            self.remaining_bits[field] = list(
-                compress(ct.tof_coefficient_table, tof_calculation_array)
+            remaining_bit_coefficients[field] = list(
+                compress(decompress_tables.tof_coefficient_table, tof_calculation_array)
             )
 
-    def _parse_binary(self):
-        """Use the TOF decoder to split up the data bits into its TOF fields."""
+        return remaining_bit_coefficients
+
+    def _parse_binary(self, case_number, tof_decoder):
+        """Use the TOF decoder to split up the data bits into its TOF fields.
+
+        The first few binary bits are only used for determining the case number
+        so either the first 5 or 6 bits need to be ignored depending the kind
+        of DE.
+        """
         # separate the binary data into its parts
-        self.parsed_bits = {}
-        # the first few bits are only used for determining the case number
-        # so either the first 5 or 6 bits need to be ignored depending the kind
-        # of DE.
-        if self.case_number in [0, 4, 6, 10, 12, 14]:
+        parsed_bits = {}
+        # check whether the first 4 or 5 bits were needed for the
+        # case determination
+        if case_number in [0, 4, 6, 10, 12, 14]:
             bit_start = 5
         else:
             bit_start = 4
         # Use the TOF decoder to chunk the data binary into its componenets
         # TOF0, TOF1, TOF2, etc.
-        for field, bit_length in self.tof_decoder.items():
-            self.parsed_bits[field] = self.DATA[bit_start : bit_start + bit_length]
+        for field, bit_length in tof_decoder._asdict().items():
+            parsed_bits[field] = self.DATA[bit_start : bit_start + bit_length]
             bit_start = bit_start + bit_length
 
-    def _decode_fields(self):
+        return parsed_bits
+
+    def _decode_fields(self, remaining_bit_coefficients, parsed_bits):
         """Use the parsed data and TOF coefficients to decode the binary."""
-        for field, tof_bits in self.parsed_bits.items():
-            needed_bits = tof_bits[-len(self.remaining_bits[field]) :]
+        for field, tof_bits in parsed_bits.items():
+            needed_bits = tof_bits[-len(remaining_bit_coefficients[field]) :]
             # Use the TOF coefficients and the bits for the current field to
             # calculate the decompressed value. Also round to 2 because the TOF
             # coefficients only have 2 decimals.
+            print(f"NEEDED BITS: { bitarray(needed_bits).tolist()}")
```

Comment:
Should these be left in? Or just used for debugging?

Reply:
Thanks for catching that! Removing those print statements